src/main/java/bio/terra/service/snapshot/flight/create/CreateSnapshotPrimaryDataQueryInterface.java

```diff
@@ -1,39 +1,77 @@
 package bio.terra.service.snapshot.flight.create;
 
+import bio.terra.common.iam.AuthenticatedUserRequest;
 import bio.terra.grammar.Query;
 import bio.terra.grammar.exception.InvalidQueryException;
+import bio.terra.model.SnapshotAccessRequestResponse;
+import bio.terra.model.SnapshotRequestContentsModel;
 import bio.terra.model.SnapshotRequestModel;
 import bio.terra.model.SnapshotRequestQueryModel;
 import bio.terra.service.common.CommonFlightUtils;
 import bio.terra.service.dataset.AssetSpecification;
 import bio.terra.service.dataset.Dataset;
 import bio.terra.service.dataset.DatasetService;
 import bio.terra.service.snapshot.Snapshot;
+import bio.terra.service.snapshot.SnapshotDao;
 import bio.terra.service.snapshot.exception.AssetNotFoundException;
+import bio.terra.service.snapshotbuilder.SnapshotBuilderService;
+import bio.terra.service.snapshotbuilder.SnapshotRequestDao;
 import bio.terra.stairway.FlightContext;
 import bio.terra.stairway.StepResult;
 import java.time.Instant;
 import java.util.List;
+import java.util.UUID;
 
 public interface CreateSnapshotPrimaryDataQueryInterface {
 
   default StepResult prepareQueryAndCreateSnapshot(
       FlightContext context,
       Snapshot snapshot,
       SnapshotRequestModel snapshotReq,
-      DatasetService datasetService)
+      DatasetService datasetService,
+      SnapshotBuilderService snapshotBuilderService,
+      SnapshotRequestDao snapshotRequestDao,
+      SnapshotDao snapshotDao,
+      AuthenticatedUserRequest userReq)
       throws InterruptedException {
-    SnapshotRequestQueryModel snapshotQuerySpec = snapshotReq.getContents().get(0).getQuerySpec();
+    AssetSpecification assetSpecification;
+    String sqlQuery;
+    Instant createdAt;
 
-    Query query = Query.parse(snapshotQuerySpec.getQuery());
-    String datasetName = query.getDatasetName();
-    Dataset dataset = datasetService.retrieveByName(datasetName);
-    AssetSpecification assetSpecification =
-        retrieveAssetSpecification(dataset, snapshotQuerySpec.getAssetName());
-    validateRootTable(query, assetSpecification);
-    String sqlQuery = translateQuery(query, dataset);
+    if (snapshotReq
+        .getContents()
+        .get(0)
+        .getMode()
+        .equals(SnapshotRequestContentsModel.ModeEnum.BYREQUESTID)) {
```

Comment:
I think if we don't pull this into its own step, did you think about making this a switch statement on `getMode`, and then pull the body of each case into a helper function? I think that might make it a lot more readable.